### PR TITLE
Add toolbutton to select action from direct entry install

### LINKED ISSUE #84: Add option to select what installer tool to use when using the direct entry
Screenshot 2024-07-22 at 10 30 01 AM

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #84

napari

Conda/pypi options are exclusive so selecting one, will leave that as the default.

To test locally this https://github.com/napari/napari-plugin-manager/pull/96/files#diff-c4a932d5522a5a713957223899c8991682affd18d838b197d3c508ea32333a1eR1215

might need a manual `if True`

### CODE REVIEW

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 1]
```diff
@@ -1178,13 +1187,34 @@ def _setup_ui(self):
...
-            trans._('install by name/url, or drop file...')
-        )
+        self.direct_entry_edit.returnPressed.connect(self._install_packages)
         self.direct_entry_edit.setVisible(visibility_direct_entry)
-        self.direct_entry_btn = QPushButton(trans._("Install"), self)
+        self.direct_entry_btn = QToolButton(self)
         self.direct_entry_btn.setVisible(visibility_direct_entry)
         self.direct_entry_btn.clicked.connect(self._install_packages)
+        self.direct_entry_btn.setText(trans._("Install"))
+        self.direct_entry_btn.setToolButtonStyle(Qt.ToolButtonTextBesideIcon)
```
Comment (dalthviz) [inline napari_plugin_manager/qt_plugin_dialog.py:1196]: Setting the popupMode of the toolbutton a behavior were you click the arrow beside the icon can be achived. So changing things here to be something like:

```suggestion
        self.direct_entry_btn.setPopupMode(QToolButton.MenuButtonPopup)
```

With that behave something like this (at least on Windows):

install_toolbutton_arrow_click

**Review by dalthviz — COMMENTED** [review]
Gave this a check on Windows and seems like things are working:

* pip installed napari case:

imagen

* conda installed napari case:

imagen

However, for the conda case, it took my a moment to realize that I had to hold the button for the menu to show up:

install_toolbutton

I think that instead of setting `self.direct_entry_btn.setToolButtonStyle(Qt.ToolButtonTextBesideIcon)` we could set the toolbutton [`popupMode`](https://doc.qt.io/qt-6/qtoolbutton.html#popupMode-prop) to `QToolButton.MenuButtonPopup`. That should enable the toolbutton to trigger the menu by clicking an arrow that gets added to the toolbutton. Left a suggestion for that in case clicking instead of holding makes more sense.

Besides that maybe disabling the toolbutton while a command is being executed could be worthy? Or maybe is expected to be able to launch multiple installs commands at the same time? 🤔

**Comment by goanpeca** [comment]
Had to make some qss changes @dalthviz, could you check again?

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 2]
```diff
@@ -1178,13 +1187,35 @@ def _setup_ui(self):
...
+        self.direct_entry_btn.setPopupMode(QToolButton.MenuButtonPopup)
```
Comment (dalthviz) [inline napari_plugin_manager/qt_plugin_dialog.py:1197]: Checking this again, I think it makes sense to move this line to be under the `IS_NAPARI_CONDA_INSTALLED` validation (around line 1216). In that way, the highlighted arrow will only be shown when there are options to select. Currently the arrow ends up showing for both the conda and pip cases.

File: napari_plugin_manager/styles.qss [hunk 3]
```diff
@@ -350,3 +350,26 @@ QtLayerList::indicator:checked {
...
+  background-color: {{ lighten(foreground, 20) }}
+}
+
+QToolButton::menu-arrow {
+  image: url("theme_{{ id }}:/down_arrow.svg");
+  top: 0px;
+  left: 0px;
+  width: 5px;
+  height: 5px;
+}
```
Comment (dalthviz) [inline napari_plugin_manager/styles.qss:375]: With this style changes (width and heigth set to 5px) on Windows I'm seeing something like this:

imagen

Also, for reference, on Windows:

* If I change the width and height to something like 8px I'm able to see something like this:

imagen

And without any custom `menu-arrow` style:

imagen

File: napari_plugin_manager/qt_plugin_dialog.py (hunk 2 above)
Comment (goanpeca) [reply napari_plugin_manager/qt_plugin_dialog.py:1197]: You are correct!

File: napari_plugin_manager/styles.qss (hunk 3 above)
Comment (goanpeca) [reply napari_plugin_manager/styles.qss:375]: Thanks for the input, however I could not make it work on Mac with bugger values, and that is why I left it "small" 5px

Comment (goanpeca) [reply napari_plugin_manager/styles.qss:375]: @dalthviz this is how it looks on mac, it seems I do not have control over the actual size of the icon nor the spacing to the sides of it :-|  on mac at least.

Screenshot 2024-08-18 at 8 32 53 PM

Comment (goanpeca) [reply napari_plugin_manager/styles.qss:375]: It seems on mac I cannot control the width of the  `MenuButtonPopup` thinguie

```
QToolButton::menu-arrow {
  image: url("theme_{{ id }}:/down_arrow.svg");
  top: 0px;
  left: 0px;
  width: 10px;
  height: 10px;
}
```

The arrow increases in size, but not the rectangular  section

Screenshot 2024-08-18 at 8 41 41 PM

Comment (dalthviz) [reply napari_plugin_manager/styles.qss:375]: Oh I see, indeed the right clickable path is quite different on macOS than on Windows 🤔 will try to give a check on macOS to see if there is something to style that part :+1:

Comment (dalthviz) [reply napari_plugin_manager/styles.qss:375]: Seems like what is needed is to use the `QToolButton[popupMode="1"]` to give the overall button extra width (only when the pop up mode `MenuButtonPopup`) and for the arrow button itself you can use to style things `QToolButton::menu-button`

For more info you can check: https://doc.qt.io/qt-5/stylesheet-examples.html#customizing-qtoolbutton

Comment (dalthviz) [reply napari_plugin_manager/styles.qss:375]: Setting as style rules something like:

```
QToolButton::menu-arrow {
  image: url("theme_{{ id }}:/down_arrow.svg");
  top: 0px;
  left: 0px;
  width: 8px;
  height: 8px;
}

QToolButton[popupMode="1"]{
padding-right: 20px;
}

QToolButton::menu-button {
 width: 16px;
}

```

I was able to get on macOS something like:

image

Comment (goanpeca) [reply napari_plugin_manager/styles.qss:375]: Awesomeeee @dalthviz indeed I did not try that!

Thanks, will implement the changes and we are good to go :)

**Pull request merged**